ucp/benchmarks/cudf_merge.py

```diff
@@ -325,30 +393,203 @@ def parse_args():
+        "specifying this argument will list the commands that should be "
+        "launched in each node. This is only a convenience function, and the "
+        "user can write the same command lines by just following the guidance "
+        "in this file's argument descriptions and existing documentation.",
+    )
     args = parser.parse_args()
-    args.devs = [int(d) for d in args.devs.split(",")]
-    args.n_chunks = len(args.devs) * args.chunks_per_dev
-    if args.n_chunks < 2:
-        raise RuntimeError(
-            f"Number of chunks must be greater than 1 (chunks-per-dev: \
-                    {args.chunks_per_dev}, devs: {args.devs})"
-        )
+
+    if args.hosts:
+        try:
+            import asyncssh  # noqa
+        except ImportError:
+            raise RuntimeError(
+                "The use of `--hosts` for SSH multi-node benchmarking requires "
+                "`asyncssh` to be installed."
+            )
+
+        if any(
+            arg
+            for arg in [
+                args.server,
+                args.num_workers,
+                args.node_idx,
+            ]
+        ):
+            raise RuntimeError(
+                "A multi-node setup using `--hosts` for automatic SSH configuration "
+                "cannot be used together with `--server`, `--num-workers` or "
+                "`--node-idx`."
+            )
+        elif args.server_file and not args.print_commands_only:
+            raise RuntimeError(
+                "Specifying `--server-file` together with `--hosts` is not "
+                "allowed, except when used with `--print-commands-only`."
+            )
+    else:
+        args.devs = [int(d) for d in args.devs.split(",")]
+        args.num_node_workers = len(args.devs) * args.chunks_per_dev
+
+        if any([args.server, args.server_file, args.server_address]):
+            if args.server_address:
+                server_host, server_port = args.server_address.split(":")
+                args.server_address = {"address": server_host, "port": int(server_port)}
+            args.server_info = args.server_file or args.server_address
+
+            if args.num_workers is None:
+                raise RuntimeError(
+                    "A multi-node setup requires specifying `--num-workers`."
+                )
+            elif args.num_workers < 2:
+                raise RuntimeError("A multi-node setup requires `--num-workers >= 2`.")
+
+            if not args.server and args.node_idx is None:
+                raise RuntimeError(
+                    "Each worker on a multi-node is required to specify `--node-num`."
+                )
+
+            args.n_chunks = args.num_workers
+        else:
+            args.n_chunks = args.num_node_workers
+
+        if args.n_chunks < 2:
+            raise RuntimeError(
+                f"Number of chunks must be greater than 1 (chunks-per-dev: \
+                        {args.chunks_per_dev}, devs: {args.devs})"
```

Comment:
This whitespace is not squashed away :(
```suggestion
                "Number of chunks must be greater than 1 (chunks-per-dev: "
                f"{args.chunks_per_dev}, devs: {args.devs})"
```

Reply:
Interesting, this was probably `black`'s doing. Probably not great at parsing this kind of string.